File: 0355-design-twitter/0355-design-twitter.java
// Method 1: Uses maxHeap to get most recent tweets
/*
# Big picture: what `getNewsFeed` must do

`getNewsFeed(userId)` needs to return up to **10 most recent tweetIds** posted by:

1. `userId` themself
2. everyone `userId` follows

“Most recent” means: newest tweet across all those users, then next newest, etc.

---

# Your data structures (what they store)

## 1) `tweetMap: Map<Integer, List<Pair>>`

For each user, you store a list of their tweets:

```java
Pair(count, tweetId)
```

* `count` is a global timestamp that increases each time any tweet is posted (`count++`).
* Because you always append, each user’s list is automatically in increasing time order:

  * oldest … newest

Example list for user 5:

```
[(0, 101), (7, 300), (12, 44)]
```

## 2) `followMap: Map<Integer, Set<Integer>>`

For each follower, the set of followees.

Example:

```
followMap.get(1) = {2, 3}
```

## 3) `Node` (heap entry)

This is the key to efficiency:

```java
static class Node {
  int time;    // tweet timestamp (Pair.count)
  int tweetId; // tweet id
  int userId;  // which user’s tweet list this came from
  int idx;     // index inside that user's list
}
```

A `Node` is like a “cursor” into one user’s tweet list.

---

# Why a heap is needed

Each user’s tweet list is individually sorted by time, but you need the **global top 10** across multiple lists.

This is exactly the classic problem:

> Merge multiple sorted lists and take the top 10 (like merging k sorted lists).

But instead of merging everything (too slow), you only ever look at the **current best candidate** from each list. The heap gives you that in `O(log F)` time, where `F` is number of sources (followees + self).

---

# `getNewsFeed` explained line-by-line

## Step 1: Max-heap by time

```java
PriorityQueue<Node> maxHeap =
    new PriorityQueue<>((a, b) -> Integer.compare(b.time, a.time));
```

* Highest `time` (most recent) comes out first.

So `poll()` always returns the newest tweet currently available among candidates.

---

## Step 2: Build the “sources” set (self + followees)

```java
Set<Integer> sources = new HashSet<>();
sources.add(userId);
sources.addAll(followMap.getOrDefault(userId, Collections.emptySet()));
```

This ensures:

* Your feed includes your own tweets even if you follow nobody.
* If the user doesn’t exist in `followMap`, it’s fine.

---

## Step 3: Initialize heap with each source’s newest tweet only

```java
for (int u : sources) {
    List<Pair> tweets = tweetMap.get(u);
    if (tweets == null || tweets.isEmpty()) continue;

    int idx = tweets.size() - 1; // newest tweet index
    Pair p = tweets.get(idx);
    maxHeap.offer(new Node(p.count, p.tweetId, u, idx));
}
```

Important insight:

* For each user, the **newest tweet** is at the end of their list.
* You push *only one* tweet per user into the heap initially.
* That means heap size ≤ number of sources.

---

## Step 4: Repeatedly take the newest, then advance within that user’s list

```java
while (ans.size() < 10 && !maxHeap.isEmpty()) {
    Node cur = maxHeap.poll();
    ans.add(cur.tweetId);

    int nextIdx = cur.idx - 1;  // next older tweet from same user
    if (nextIdx >= 0) {
        List<Pair> tweets = tweetMap.get(cur.userId);
        Pair p = tweets.get(nextIdx);
        maxHeap.offer(new Node(p.count, p.tweetId, cur.userId, nextIdx));
    }
}
```

This is the k-way merge logic:

* You pop the newest tweet overall.
* If that tweet came from user U at index `idx`, then the next candidate from user U is their tweet at `idx - 1` (the next older one).
* Push that next older tweet into the heap.

This way, you always keep the heap filled with “current best remaining tweet” from each user.

---

# Thorough example walkthrough (heap contents every step)

Let’s simulate:

### Actions

1. `postTweet(1, 101)`
2. `postTweet(2, 201)`
3. `postTweet(1, 102)`
4. `postTweet(2, 202)`
5. `postTweet(3, 301)`
6. `follow(1, 2)`
7. `follow(1, 3)`
8. `getNewsFeed(1)`

### Timestamps assigned (`count++`)

Assume `count` starts at 0:

* time 0: user1 tweet 101  → Pair(0,101)
* time 1: user2 tweet 201  → Pair(1,201)
* time 2: user1 tweet 102  → Pair(2,102)
* time 3: user2 tweet 202  → Pair(3,202)
* time 4: user3 tweet 301  → Pair(4,301)

So `tweetMap` looks like:

* user 1 list: `[(0,101), (2,102)]`
* user 2 list: `[(1,201), (3,202)]`
* user 3 list: `[(4,301)]`

And `followMap.get(1) = {2, 3}`

---

## Now call: `getNewsFeed(1)`

### Step A: Build sources

`sources = {1, 2, 3}`

### Step B: Push newest tweet from each source into heap

* For user 1: newest idx=1 → (time=2, tweet=102)

  * push Node(2,102,user1,idx1)

* For user 2: newest idx=1 → (time=3, tweet=202)

  * push Node(3,202,user2,idx1)

* For user 3: newest idx=0 → (time=4, tweet=301)

  * push Node(4,301,user3,idx0)

Heap (conceptually ordered by time):

* top: (4,301,u3,i0)
* then: (3,202,u2,i1)
* then: (2,102,u1,i1)

`ans = []`

---

### Iteration 1

Pop heap → (4,301,u3,i0)

* ans = [301]
  Push next older from u3:
* nextIdx = 0-1 = -1 → none

Heap now:

* (3,202,u2,i1)
* (2,102,u1,i1)

---

### Iteration 2

Pop → (3,202,u2,i1)

* ans = [301, 202]
  Push next older from u2:
* nextIdx = 1-1 = 0 → tweetMap[2][0] = (1,201)
* push Node(1,201,u2,i0)

Heap now:

* (2,102,u1,i1)
* (1,201,u2,i0)

---

### Iteration 3

Pop → (2,102,u1,i1)

* ans = [301, 202, 102]
  Push next older from u1:
* nextIdx = 1-1 = 0 → tweetMap[1][0] = (0,101)
* push Node(0,101,u1,i0)

Heap now:

* (1,201,u2,i0)
* (0,101,u1,i0)

---

### Iteration 4

Pop → (1,201,u2,i0)

* ans = [301, 202, 102, 201]
  Push next older from u2:
* nextIdx = 0-1 = -1 → none

Heap now:

* (0,101,u1,i0)

---

### Iteration 5

Pop → (0,101,u1,i0)

* ans = [301, 202, 102, 201, 101]
  Push next older from u1:
* nextIdx = -1 → none

Heap empty → stop.

✅ Final result: `[301, 202, 102, 201, 101]`

That is exactly the 5 tweets in global descending time:
time 4,3,2,1,0

---

# Why this is efficient

Let:

* `F` = number of sources (followees + user)

Initialization pushes ≤ F nodes.

Then you do ≤ 10 iterations:

* each iteration is one `poll()` and at most one `offer()`
* each heap op costs `O(log F)`

So:

* Time for `getNewsFeed`: about `O(F log F + 10 log F)`
  (dominated by initial pushing of at most F tweets)
* Space: `O(F)` heap + output list

This is far better than:

* dumping all tweets from all followees into one list and sorting (could be huge).

---

# A couple of subtle correctness notes

1. **Ties in timestamp**
   Your timestamps are strictly increasing (`count++`), so no ties. Great.

2. **User follows themselves**
   Even if user 1 follows user 1, using a `Set` prevents duplicates.

3. **Order among the returned tweets**
   The returned list is correct ordering by time, but the code does not try to break ties by tweetId (not needed).
*/

class Twitter {

    static class Pair{
        int count;
        int tweetId;
        Pair(int count, int tweetId){
            this.count = count;
            this.tweetId = tweetId;
        }
    }

    static class Node{
        int time;
        int tweetId;
        int userId;
        int idx;

        Node(int time, int tweetId, int userId, int idx){
            this.time = time;
            this.tweetId = tweetId;
            this.userId = userId;
            this.idx = idx;
        }
    }
    
    Map<Integer, Set<Integer>> followMap;
    Map<Integer, List<Pair>> tweetMap;
    int count;

    public Twitter() {
        followMap = new HashMap<>();
        tweetMap = new HashMap<>();
        count = 0;
    }
    
    public void postTweet(int userId, int tweetId) {
        tweetMap.computeIfAbsent(userId, k -> new ArrayList<>()).add(new Pair(count++, tweetId));
    }
    
    public List<Integer> getNewsFeed(int userId) {

        // Max-heap by timestamp (most recent first)
        PriorityQueue<Node> maxHeap = new PriorityQueue<>((a , b) -> Integer.compare(b.time, a.time));

        // sources = userId + everyone they follow
        Set<Integer> sources = new HashSet<>();
        sources.add(userId);
        sources.addAll(followMap.getOrDefault(userId, Collections.emptySet()));

        // Push each source's most recent tweet into heap
        for(int u: sources){
            List<Pair> tweets = tweetMap.get(u);
            if(tweets == null || tweets.isEmpty()){
                continue;
            }

            int idx = tweets.size() - 1;
            Pair p = tweets.get(idx);
            maxHeap.offer(new Node(p.count, p.tweetId, u, idx));
        }

        // Pull up to 10 most recent tweets
        List<Integer> ans = new ArrayList<>(10);
        while(ans.size() < 10 && !maxHeap.isEmpty()){
            Node cur = maxHeap.poll();
            ans.add(cur.tweetId);


            // Push next older tweet from the same user (if any)
            int prevIdx = cur.idx - 1;
            if(prevIdx >= 0){
                List<Pair> tweets = tweetMap.get(cur.userId);
                Pair p = tweets.get(prevIdx);
                maxHeap.offer(new Node(p.count, p.tweetId, cur.userId, prevIdx));
            }
        }

        return ans;
    }
    
    public void follow(int followerId, int followeeId) {
        followMap.computeIfAbsent(followerId, k -> new HashSet<>()).add(followeeId);
    }
    
    public void unfollow(int followerId, int followeeId) {
        // followMap.getOrDefault(followerId, Collections.emptySet()).remove(followeeId);
        // followMap.computeIfPresent(followerId, (k, set) -> {
        //     set.remove(followeeId);
        //     return set; 
        // });

        followMap.computeIfPresent(followerId, (k, set) -> set.remove(followeeId) ? set : set);
    }
}

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter obj = new Twitter();
 * obj.postTweet(userId,tweetId);
 * List<Integer> param_2 = obj.getNewsFeed(userId);
 * obj.follow(followerId,followeeId);
 * obj.unfollow(followerId,followeeId);
 */

File: 0355-design-twitter/0355-design-twitter.cpp
// Method 1: MaxHeap for getting news feed
/*
### Hint 1

Identify the operations you must support:

```cpp
postTweet(userId, tweetId)
getNewsFeed(userId)
follow(followerId, followeeId)
unfollow(followerId, followeeId)
```

The hardest operation is `getNewsFeed`.

---

### Hint 2

For each user, store the users they follow:

```cpp
unordered_map<int, unordered_set<int>> following;
```

Also treat a user as following themselves when building their feed, so their own tweets appear.

---

### Hint 3

Store each tweet with a timestamp so you can compare recency:

```cpp
struct Tweet {
    int tweetId;
    int time;
};
```

Maintain a global increasing counter:

```cpp
int timestamp = 0;
```

When posting:

```cpp
tweets[userId].push_back({tweetId, timestamp++});
```

---

### Hint 4

For each user, keep their tweets in chronological order:

```cpp
unordered_map<int, vector<Tweet>> tweets;
```

Since new tweets are appended, the most recent tweet is at the end of the vector.

---

### Hint 5

To build the news feed, you need the 10 newest tweets among several already-sorted tweet lists.

This is similar to merging `k` sorted lists.

Use a **max-heap** containing the most recent tweet from each relevant user.

---

### Hint 6

Each heap entry needs enough information to find the next older tweet from the same user:

```cpp
struct Entry {
    int time;
    int tweetId;
    int userId;
    int index;
};
```

Here, `index` is the tweet’s position in that user’s tweet vector.

---

### Hint 7

Initially, push the latest tweet from:

* the requesting user
* every user they follow

Then repeat at most 10 times:

1. Pop the newest tweet.
2. Add its `tweetId` to the result.
3. Push the previous tweet from the same user, if one exists.

---

### Hint 8

The core feed logic looks like:

```cpp
while (!maxHeap.empty() && feed.size() < 10) {
    auto current = maxHeap.top();
    maxHeap.pop();

    feed.push_back(current.tweetId);

    if (current.index > 0) {
        // Push the previous tweet from current.userId
    }
}
```

---

### Hint 9

For following:

```cpp
following[followerId].insert(followeeId);
```

For unfollowing:

```cpp
following[followerId].erase(followeeId);
```

Be careful with self-following. An easy design is to never store self-following and simply include the user separately in `getNewsFeed`.

---

### Suggested class state

```cpp
class Twitter {
private:
    int timestamp;

    unordered_map<int, vector<pair<int, int>>> tweets;
    unordered_map<int, unordered_set<int>> following;
};
```

You could store each tweet as:

```cpp
{timestamp, tweetId}
```

---

### Complexity

Let `f` be the number of followed users.

* `postTweet`: `O(1)`
* `follow`: average `O(1)`
* `unfollow`: average `O(1)`
* `getNewsFeed`: approximately `O((f + 10) log f)`

The key insight is that you should not combine and sort every tweet. Only keep one candidate tweet per user in the heap at a time.
*/
class Twitter {
private:
    unordered_map<int, unordered_set<int>> following;
    struct Tweet{
        int tweetId;
        int time;
    };
    int timestamp = 0;
    unordered_map<int, vector<Tweet>> tweets;
    struct Entry{
        int time;
        int tweetId;
        int userId;
        int index;
    };

    // Makes the entry with the greatest timestamp appear at the top of the priority queue.
    struct Compare{
        bool operator()(const Entry& a, const Entry& b) const {
            return a.time < b.time;
        }
    };

public:
    Twitter() {}
    
    void postTweet(int userId, int tweetId) {
        tweets[userId].push_back({tweetId, timestamp++});
    }
    
    vector<int> getNewsFeed(int userId) {
        vector<int> feed;
        priority_queue<Entry, vector<Entry>, Compare> maxHeap;

        // Adding user's latest tweet
        if(!tweets[userId].empty()){
            int index = static_cast<int>(tweets[userId].size()) - 1;
            Tweet& tweet = tweets[userId][index];

            maxHeap.push({tweet.time, tweet.tweetId, userId, index});
        }

        // Adding the latest tweet from each followed user
        for(int followeeId : following[userId]){
            if(followeeId == userId || tweets[followeeId].empty()){
                continue;
            }

            int index = static_cast<int> (tweets[followeeId].size()) - 1;
            Tweet& tweet = tweets[followeeId][index];

            maxHeap.push({tweet.time, tweet.tweetId, followeeId, index});
        }

        while(!maxHeap.empty() && feed.size() < 10){
            Entry current = maxHeap.top();
            maxHeap.pop();

            feed.push_back(current.tweetId);

            // Push the next older tweet from the same user
            int previousIndex = current.index - 1;

            if(previousIndex >= 0){
                Tweet& previousTweet = tweets[current.userId][previousIndex];

                maxHeap.push({previousTweet.time, previousTweet.tweetId, current.userId, previousIndex});
            }
        }

        return feed;
    }
    
    void follow(int followerId, int followeeId) {
        if(followerId != followeeId){
            following[followerId].insert(followeeId);
        }
    }
    
    void unfollow(int followerId, int followeeId) {
        following[followerId].erase(followeeId);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna